src/alfasim_sdk/result_reader/aggregator.py

```diff
@@ -0,0 +1,1344 @@
+        directory = filename.parent
+        old_directory = os.getcwd()
+        try:
+            os.chdir(directory)
+            try:
+                return h5py.File(filename.name, "r", libver="latest", swmr=True)
+            except OSError as os_error:
+                swmr_message = (
+                    "Unable to open file (file is not already open for SWMR writing)"
+                )
+                if str(os_error) == swmr_message:
+                    return h5py.File(filename.name, "r", libver="latest", swmr=False)
+                raise
+
+        except PermissionError:
+            raise PermissionError(
+                f"Could not access folder {filename}.\n"
+                f'If you are using your local "Downloads" directory, consider changing the project file\n'
+                f"to somewhere else as some services try to synchronize that directory and interfere with\n"
+                f"the simulation.\n"
+                "\n"
+                f'1 - Click in "Save as".\n'
+                f"2 - Select a different path, outside Download folder.\n"
+                f"3 - Run simulation again.\n"
+            )
+        finally:
+            os.chdir(old_directory)
+
+    # When a new result file is created its metadata contents are not complete, and the file
+    # has not been put into SWMR mode yet (SWMR mode does not allow new groups, attributes,
+    # and/or data sets to be created).
+    # Since it is not possible to open files in SWMR mode for reading if a writer is not
+    # present fall back to open files from former runs in normal read mode.
+    # Since a new file (with incomplete metadata) is open by the writer in normal mode we
+    # need avoid reading that incomplete file.
+    # To avoid race conditions a file `<result_file_name>.creating` is create in the same
+    # folder before the actual result file is created. After the metadata is writen and the
+    # writer enters SWMR mode the `<result_file_name>.creating` is removed to signal readers
+    # that is safe to read that file.
+    files = set(result_directory.glob(RESULT_FILE_PREFIX + "*"))
+    result_files = set()
+    files_under_creation = set()
+    for f in files:
+        ext = f.suffix
+        if ext == ".creating":
+            # Take note of any file to ignore.
+            name, _ = os.path.splitext(f)
+            files_under_creation.add(Path(name))
+        elif ext == "":
+            result_files.add(f)
+        else:
+            raise ValueError(f"Unknown ext: {ext}")
+
+    prefix_len = len(RESULT_FILE_PREFIX)
+    result_files.difference_update(files_under_creation)  # Ignore incomplete files.
+    result_files_sorted = (
+        (int(filename.name[prefix_len:]), open_result_file(filename))
+        for filename in result_files
+    )
+    result_files_sorted_dict = _Dict(sorted(result_files_sorted, key=lambda x: x[0]))
+
+    # Close the files once the container is collected.
+    import weakref
+
+    weakref.finalize(
```

Comment:
I see what you did here, but perhaps it would be better to make this a context manager, which closes the files at the end.

Using weakrefs for this is a bit implicit/magical, so I would prefer a more explicit approach if possible.

Reply:
That was a bad decision when I wrote it...
I will check this.

Reply:
Gentle ping here